File: application/backend-credit-card-enrollment/backend-c#/Common/Reaction/ReactionController.cs
using CreditCardEnrollment.Common.Ambar;
using CreditCardEnrollment.Common.EventStore;
using CreditCardEnrollment.Common.Projection;
using CreditCardEnrollment.Common.SerializedEvent;

namespace CreditCardEnrollment.Common.Reaction;

public abstract class ReactionController {
    private readonly PostgresTransactionalEventStore _postgresTransactionalEventStore;
    private readonly MongoTransactionalProjectionOperator _mongoTransactionalProjectionOperator;
    private readonly Deserializer _deserializer;
    private readonly ILogger<ReactionController> _logger;

    protected ReactionController(
        PostgresTransactionalEventStore postgresTransactionalEventStore,
        MongoTransactionalProjectionOperator mongoTransactionalProjectionOperator,
        Deserializer deserializer,
        ILogger<ReactionController> logger
    ) {
        _postgresTransactionalEventStore = postgresTransactionalEventStore;
        _mongoTransactionalProjectionOperator = mongoTransactionalProjectionOperator;
        _deserializer = deserializer;
        _logger = logger;
    }

    protected string ProcessReactionHttpRequest(AmbarHttpRequest ambarHttpRequest, ReactionHandler reactionHandler) {
        try {
            _logger.LogDebug(
                "Starting to process reaction for event name: {EventName} using handler: {HandlerName}", 
                ambarHttpRequest.SerializedEvent.EventName,
                reactionHandler.GetType().Name
            );
            _postgresTransactionalEventStore.BeginTransaction();
            _mongoTransactionalProjectionOperator.StartTransaction();
            reactionHandler.React(_deserializer.Deserialize(ambarHttpRequest.SerializedEvent));
            _postgresTransactionalEventStore.CommitTransaction();
            _mongoTransactionalProjectionOperator.CommitTransaction();

            _postgresTransactionalEventStore.AbortDanglingTransactionsAndReturnConnectionToPool();
            _mongoTransactionalProjectionOperator.AbortDanglingTransactionsAndReturnSessionToPool();

            _logger.LogDebug(
                "Reaction successfully processed for event name: {EventName} using handler: {HandlerName}", 
                ambarHttpRequest.SerializedEvent.EventName,
                reactionHandler.GetType().Name
            );
            return AmbarResponseFactory.SuccessResponse();
        } catch (Exception ex) when (ex.Message?.StartsWith("Unknown event type") == true) {
            _postgresTransactionalEventStore.AbortDanglingTransactionsAndReturnConnectionToPool();
            _mongoTransactionalProjectionOperator.AbortDanglingTransactionsAndReturnSessionToPool();
            return AmbarResponseFactory.SuccessResponse();
        } catch (Exception ex) {
            _postgresTransactionalEventStore.AbortDanglingTransactionsAndReturnConnectionToPool();
            _mongoTransactionalProjectionOperator.AbortDanglingTransactionsAndReturnSessionToPool();
            _logger.LogError("Exception in ProcessReactionHttpRequest: {0}, {1}", ex.Message, ex.StackTrace);
            return AmbarResponseFactory.RetryResponse(ex);
        }
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/common/reaction/ReactionController.ts
import { AmbarHttpRequest } from '../ambar/AmbarHttpRequest';
import { AmbarResponseFactory } from '../ambar/AmbarResponseFactory';
import { PostgresTransactionalEventStore } from '../eventStore/PostgresTransactionalEventStore';
import { MongoTransactionalProjectionOperator } from '../projection/MongoTransactionalProjectionOperator';
import { Deserializer } from '../serializedEvent/Deserializer';
import { log } from '../util/Logger';
import { ReactionHandler } from './ReactionHandler';

export abstract class ReactionController {
    constructor(
        private readonly postgresTransactionalEventStore: PostgresTransactionalEventStore,
        private readonly mongoTransactionalProjectionOperator: MongoTransactionalProjectionOperator,
        private readonly deserializer: Deserializer
    ) {}

    protected async processReactionHttpRequest(
        ambarHttpRequest: AmbarHttpRequest,
        reactionHandler: ReactionHandler
    ): Promise<string> {
        try {
            log.debug(
                `Starting to process reaction for event name: ${ambarHttpRequest.payload.event_name} using handler: ${reactionHandler.constructor.name}`
            );
            await this.postgresTransactionalEventStore.beginTransaction();
            await this.mongoTransactionalProjectionOperator.startTransaction();
            await reactionHandler.react(this.deserializer.deserialize(ambarHttpRequest.payload));
            await this.postgresTransactionalEventStore.commitTransaction();
            await this.mongoTransactionalProjectionOperator.commitTransaction();

            await this.postgresTransactionalEventStore.abortDanglingTransactionsAndReturnConnectionToPool();
            await this.mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();

            log.debug(
                `Reaction successfully processed for event name: ${ambarHttpRequest.payload.event_name} using handler: ${reactionHandler.constructor.name}`
            );
            return AmbarResponseFactory.successResponse();
        } catch (error) {
            if (error instanceof Error && error.message.startsWith('Unknown event type')) {
                await this.postgresTransactionalEventStore.abortDanglingTransactionsAndReturnConnectionToPool();
                await this.mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();
                log.debug(
                    `Unknown event in reaction ignored for event name: ${ambarHttpRequest.payload.event_name} using handler: ${reactionHandler.constructor.name}`
                );
                return AmbarResponseFactory.successResponse();
            }

            await this.postgresTransactionalEventStore.abortDanglingTransactionsAndReturnConnectionToPool();
            await this.mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();
            log.error('Exception in ProcessReactionHttpRequest:', error as Error);
            return AmbarResponseFactory.retryResponse(error as Error);
        }
    }
}